### PR TITLE
CLOUDSTACK-8239 Add VirtIO SCSI support for KVM hosts

### PR DESCRIPTION
**Pull request by nathanejohnson**
This is an initial implementation that we have working in the lab.

devmanagerscreenshot
spanned_sceenshot

[win_virtio_scsi.xml.zip](https://github.com/apache/cloudstack/files/788014/win_virtio_scsi.xml.zip)

@wido @syed Would love to get some feedback on this approach.

### CODE REVIEW

**Comment by kiwiflyer** [comment]
@remibergsma @ustcweizhou This might be of interest to you guys as well.

**Comment by remibergsma** [comment]
Nice work @kiwiflyer @dmabry, thanks! Will give it a go soon.

**Comment by wido** [comment]
Very nice indeed! I will take a look asap.

**Comment by dmabry** [comment]
We are deploying this to our QA environment right now and hope to have it tested in a few days.  Great work @kiwiflyer and @nathanejohnson.

**Comment by borisstoyanov** [comment]
Thanks for this PR @nathanejohnson, I will build it and run smoketests agains it. 
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
@nathanejohnson Thanks, for this good enhancement. Can we have a marvin tests that covers the new functionality?

File: plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java [hunk 1]
```diff
@@ -2059,6 +2060,13 @@ So if getMinSpeed() returns null we fall back to getSpeed().
         final InputDef input = new InputDef("tablet", "usb");
         devices.addDevice(input);
 
+        // If we're using virtio scsi, then we need to add a virtual scsi controller
+        if (getGuestDiskModel(vmTO.getPlatformEmulator()) == DiskDef.DiskBus.SCSI) {
+            final SCSIDef sd = new SCSIDef((short)0, 0, 0, 9, 0);
+            devices.addDevice(sd);
+            s_logger.debug("Adding scsi definition:\n" + sd.toString());
```
Comment (wido) [inline plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java]: Might want to include the name of the VM here as well? And SCSI is written in upper-case.

File: plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java [hunk 2]
```diff
@@ -2158,6 +2166,7 @@ public int compare(final DiskTO arg0, final DiskTO arg1) {
 
             if (diskBusType == null) {
                 diskBusType = getGuestDiskModel(vmSpec.getPlatformEmulator());
+                s_logger.debug("disk bus type derived from getPlatformEmulator: " + vmSpec.getPlatformEmulator() + ", diskbustype is: "+diskBusType.toString());
```
Comment (wido) [inline plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java]: Same here, please add the name of the VM

File: plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java [hunk 3]
```diff
@@ -2216,6 +2225,8 @@ public int compare(final DiskTO arg0, final DiskTO arg1) {
                     disk.setCacheMode(DiskDef.DiskCacheMode.valueOf(volumeObjectTO.getCacheMode().toString().toUpperCase()));
                 }
             }
+
+            s_logger.debug("Adding disk: " + disk.toString());
```
Comment (wido) [inline plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java]: Same here, please include the name of the VM

Also, do we really need all these logging lines?

**Comment by nathanejohnson** [comment]
As an update, I'm working on a Marvin test, and I'm also planning to address wido's concerns.  Today had many distractions unfortunately.

**Comment by borisstoyanov** [comment]
@nathanejohnson thank you, let me know when you're done with the changes so I could kick in smoke tests.

**Comment by nathanejohnson** [comment]
@wido I have removed the extraneous log messages.

@boris I have added a smoke test for this that tests both virsh output from the host and lspci / lsblk output from the guest.  Here is the output from me running this in a bubble:

Test that libvirt properly created domain with scsi controller ... === TestName: test_01_verify_libvirt | Status : SUCCESS ===
ok
Verify that libvirt settings are as expected after a VM stop / start ... === TestName: test_02_verify_libvirt_after_restart | Status : SUCCESS ===
ok
Verify that libvirt settings are expected after a disk add ... === TestName: test_03_verify_libvirt_attach_disk | Status : SUCCESS ===
ok
Verify that guest sees scsi controller and disks ... === TestName: test_04_verify_guest_lspci | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 4 tests in 658.770s

OK

**Comment by borisstoyanov** [comment]
Thanks @nathanejohnson, lets run that with BO
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by nathanejohnson** [comment]
@borisstoyanov Quick questions, what tags need to be present to get picked up by BO?  I'm worried my smoke test might get skipped.

**Comment by nathanejohnson** [comment]
@borisstoyanov so the good news is it didn't skip the test.  the bad news is it didn't seem to have proper credentials for ssh'ing into your hosts.  It did pass the test where it ssh'ed into the guest, however.  So this at least demonstrates a guest is happy.

Do you have any tips on ways of making the smoke test play nice with BO wrt ssh'ing into the kvm hosts?

**Comment by borisstoyanov** [comment]
Thanks @nathanejohnson I will have a look

**Comment by remibergsma** [comment]
Tested it manually, works great. LGTM.

**Comment by borisstoyanov** [comment]
I think that's the issue at our environment, it's trying to ssh via hostname instead of IP. Can you change that to IP please @nathanejohnson ? 
`====Trying SSH Connection: Host:pr1955-t929-kvm-centos6-kvm1 User:root                                   Port:22 RetryCnt:3===`

**Comment by nathanejohnson** [comment]
@borisstoyanov I have reworked the test so that it should now use the host IP versus the hostname.  Could you please re-run the blue orangutan test?  Thank you!

**Comment by nathanejohnson** [comment]
Also, I just made a commit, I have not squashed it because I wanted to get some feedback.  I noticed that vmware uses details to pass rootDiskController for the bus type, and there was even some code that inspected this in the kvm plugin.  With this latest commit, it will actually honor either OS type "Other PV Virtio-SCSI" or rootDiskController passed in via details, with details overriding OS type.  I also made a minor tweak to the UI to allow setting the root disk bus via the add template dialog in the UI:

template_screen

So this way you can select virtio scsi by one of the following three ways:

1) selecting os type as "Other PV Virtio-SCSI (64 bit)"
2) uploading a template and specifying details[0].rootDiskController=scsi, either from the UI or the API, then creating a VM from this template
3) deploying a VM and specifying details[0].rootDiskController=scsi , currently only available via the API

**Comment by nathanejohnson** [comment]
@borisstoyanov Whenever you get a second could you kick off a blue orangutan smoke test?  I believe I have fixed the issue with ssh'ing into the KVM hosts, I sue the IP now instead of hostname.

**Comment by borisstoyanov** [comment]
@nathanejohnson sure, sorry I've missed your comment about the test fix earlier. 
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by kiwiflyer** [comment]
tag:mergeready

**Comment by wido** [comment]
I would like to give a LGTM, but it's only the logging lines which I don't like yet.

They seem like personal debug lines, but a admin will never know what they actually mean.

In addition, I would like to see discard/unmap enabled by default for Virtio-SCSI disks: https://blog.widodh.nl/2015/12/using-trimdiscard-with-ceph-rbd-and-libvirt/

This will also allow for QCOW2 and RBD trimming.

**Comment by nathanejohnson** [comment]
@wido I have removed all of the debug logging statements that didn't exist before I touched the code, I have also added the discard option to the disks.

**Comment by wido** [comment]
Do we want to pass the unmap not as a string? Shouldn't that be a enum?

**Comment by nathanejohnson** [comment]
@wido Actually I was just about to push a commit that changed the string to enum, but builds are failing locally because of this test:
    
    assertTrue(QuotaAlertManagerImpl.getDifferenceDays(now, new DateTime(now).plusDays(1).toDate()) == 1L);

Daylight savings time strikes again

**Comment by wido** [comment]
@nathanejohnson: Good! I would like to see that being enums.

**Comment by nathanejohnson** [comment]
I just re-depoyed and re-ran the test:

    Test that libvirt properly created domain with scsi controller ... === TestName: test_01_verify_libvirt | Status : SUCCESS ===
    ok
    Verify that libvirt settings are as expected after a VM stop / start ... === TestName: test_02_verify_libvirt_after_restart | Status : SUCCESS ===
    ok
    Verify that libvirt settings are expected after a disk add ... === TestName: test_03_verify_libvirt_attach_disk | Status : SUCCESS ===
    ok
    Verify that guest sees scsi controller and disks ... === TestName: test_04_verify_guest_lspci | Status : SUCCESS ===
    ok
    Update os type to Ubuntu, change vm details rootdiskController ... === TestName: test_05_change_vm_ostype_restart | Status : SUCCESS ===
    ok
    Verify that guest sees scsi controller and disks after switching ostype and rdc ... === TestName: test_06_verify_guest_lspci_again | Status : SUCCESS ===
    ok
    
    ----------------------------------------------------------------------
    Ran 6 tests in 785.914s
    
    OK

@wido , I believe all of your concerns should be addressed.  trim is enabled, it's using an enum and not a string, and spurious log messages have been eliminated.

I have squashed all of the commits into one.  I'm at a point where I'm happy with the code.

**Comment by wido** [comment]
LGTM from the code

**Comment by kiwiflyer** [comment]
@karuturi  3 x LGTM, testing successful. Ready for Merge.

**Comment by wido** [comment]
Yes, we are indeed ready for a merge. Shall we do that?

**Comment by karuturi** [comment]
merging..

**Comment by DaanHoogland** [comment]
@nathanejohnson @wido @remibergsma @karuturi @kiwiflyer @dmambry
in #2044 tests are failing on vmware while this is only for kvm. Can we add a check on the tests for this or is it supposed to work for all hypervisors?

**Comment by nathanejohnson** [comment]
@DaanHoogland @wido @remibergsma @karuturi @kiwiflyer I think I see the issue with the test, I should hopefully have a new PR opened later today.

**Pull request merged**